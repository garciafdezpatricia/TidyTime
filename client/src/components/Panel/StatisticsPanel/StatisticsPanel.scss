@use "../../../../styles/variables" as *;

.statistics {
    height: 80%;
    width: 100%;
    display: grid;
    justify-items: center;
    align-items: center;
    grid-template-columns: repeat(2, 1fr);
    grid-template-rows: auto 1fr;
    column-gap: 1rem;

    @media screen and (min-width: 320px) and (max-width: 768px) {
        display: flex;
        flex-direction: column;
        gap: 2rem;
    }

    .task-progress {
        box-shadow: 0px 4px 20px 0px rgba(0,0,0,0.4);
        width: 100%;
        height: 100%;
        padding: 2rem 4rem;
        box-sizing: border-box;
        background-color: $white;
        border-radius: .4rem;
        display: flex;
        flex-direction: column;
        align-items: center;
        gap: .5rem;
        grid-column: 2/2;
        grid-row: 1/1;
        svg {
            width: 35%;
            height: auto;
        }

        @media screen and (min-width: 320px) and (max-width: 768px) {
            padding: 1rem;
            font-size: .8rem;
        }

        @media screen and (min-width: 768px) and (max-width: 1200px) {
            padding: 2rem 1rem;
        }
    }

    .upcoming-events {
        box-shadow: 0px 4px 20px 0px rgba(0,0,0,0.4);
        display: flex;
        flex-direction: column;
        gap: .5rem;
        border-radius: .4rem;
        background-color: $white;
        height: 100%;
        width: 100%;
        grid-column: 1/1;
        grid-row: 1/-1;
        padding: 2rem 4rem;
        box-sizing: border-box;
        text-align: center;

        @media screen and (min-width: 320px) and (max-width: 768px) {
            padding: 1rem;
            font-size: .8rem;
        }

        @media screen and (min-width: 768px) and (max-width: 1200px) {
            padding: 2rem 1rem;
        }

        .event-list {
            height: 100%;
            overflow-y: auto;
            display: flex;
            align-items: center;
            flex-direction: column;
            gap: .3rem;
        }

        .event {
            width: 80%;
            display: flex;
            flex-direction: column;
            justify-content: center;
            align-items: center;
            background-color: $hover-green;
            padding: .3rem;
            border-radius: .4rem;
            line-height: 1.7rem;

            .event-title {
                font-weight: bold;
            }
        }
    }
}

.empty-events {
    color: $medium-grey;
    width: 100%;
    height: 100%;
    display: flex;
    align-items: center;
    justify-content: center;
}