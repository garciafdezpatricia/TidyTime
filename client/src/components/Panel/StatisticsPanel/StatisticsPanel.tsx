import { Event } from "@/src/model/Scheme";
import { useEventContext } from "../../Context/EventContext";
import { useTaskContext } from "../../Context/TaskContext";
import DonutChart from "../../DonutChart/DonutChart"
import { useEffect, useState } from "react";

export default function StatisticsPanel() {

    const { tasks } = useTaskContext();
    const { events } = useEventContext();

    const [eventsToShow, setEventsToShow] = useState<Event[]>([])
    const [totalTasks, setTotalTasks] = useState(0);
    const [doneTasks, setDoneTasks] = useState(0);

    useEffect(() => {
      const today = new Date();
      today.setHours(0, 0, 0, 0);
      const upcomingEvents = events.filter((event) => {
        const eventDate = event.start;
        eventDate.setHours(0, 0, 0, 0);
        return eventDate >= today && daysLeft(eventDate) <= 20
      });
      setEventsToShow(upcomingEvents);
    }, [events])

    const daysLeft = (fechaDada:Date) => {
        const currentDate = new Date();
        const miliseconds = fechaDada.getTime() - currentDate.getTime();
        return Math.ceil(miliseconds / (1000 * 60 * 60 * 24));
    }

    useEffect(() => {
      if (tasks && tasks.length > 0) {
        let totalCounter = 0;
        let doneCounter = 0;
        tasks.forEach((list) => {
          if (list.value) {
            totalCounter += list.value.length;
            doneCounter += list.value.filter((task) => task.done).length;
          }
        })
        setTotalTasks(totalCounter);
        setDoneTasks(doneCounter);
      } else {
        setTotalTasks(0);
        setDoneTasks(0)
      }
    }, [tasks])

    return (
    <article className="statistics">
        <article className="task-progress">
          <h2>Completed tasks</h2>
          <DonutChart total={totalTasks} value={doneTasks} />
        </article>
        <article className="upcoming-events">
          <h2>Upcoming events</h2>
          <div className="event-list">
              {
                eventsToShow.length > 0 
                  ? eventsToShow.sort((a:any, b:any) => a.start - b.start).map((event, index) => {
                    const days = daysLeft(event.start)
                    return (
                      <section className="event" key={index}>
                          <p className="event-title">{event.title.toUpperCase()}</p>
                          <p style={{textDecoration: 'underline'}}>{days === 0 ? "Today" : days >= 2 ? `${days} days left` : `${days} day left`}</p>
                          <p>{event.start.toLocaleDateString()} - {event.end.toLocaleDateString()}</p>
                      </section>
                    )
                  })
                  : <p className="empty-events">No upcoming events!</p>
              }
          </div>
        </article>
      </article>
    )
}