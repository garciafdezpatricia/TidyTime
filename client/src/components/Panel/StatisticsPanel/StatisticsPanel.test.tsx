import { render } from "@testing-library/react";
import '@testing-library/jest-dom';
import StatisticsPanel from '@/src/components/Panel/StatisticsPanel/StatisticsPanel';

describe('StatisticsPanel', () => {
    it('renders without crashing', () => {
      const { container } = render(<StatisticsPanel />);
      expect(container).toBeInTheDocument();
    });

    it('matches snapshot', () => {
        const { asFragment } = render(<StatisticsPanel />);
        expect(asFragment()).toMatchSnapshot();
      });
  });