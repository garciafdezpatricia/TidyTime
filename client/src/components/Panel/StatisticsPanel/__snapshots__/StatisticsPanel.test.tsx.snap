// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatisticsPanel matches snapshot 1`] = `
<DocumentFragment>
  <article
    class="statistics"
  >
    <article
      class="task-progress"
      data-testid="task-progress"
    >
      <h2>
        home.taskStat
      </h2>
      <svg
        height="20%"
        viewBox="0 0 120 120"
        width="20%"
      >
        <circle
          cx="60"
          cy="60"
          fill="transparent"
          r="50"
          stroke="#d8d8d8"
          stroke-width="10"
        />
        <circle
          cx="60"
          cy="60"
          fill="transparent"
          r="50"
          stroke="#000000"
          stroke-dasharray="314.1592653589793"
          stroke-dashoffset="0"
          stroke-width="10"
          style="transition: stroke-dashoffset 0.3s ease-out;"
          transform="rotate(-90 60 60)"
        />
        <text
          data-testid="total-tasks-donut"
          dy=".3em"
          font-size="20"
          text-anchor="middle"
          x="60"
          y="60"
        >
          0/0
        </text>
      </svg>
    </article>
    <article
      class="upcoming-events"
      data-testid="upcoming-events"
    >
      <h2>
        home.upcomingEvents.title
      </h2>
      <div
        class="event-list"
      >
        <p
          class="empty-events"
          data-testid="empty-events"
        >
          home.upcomingEvents.emptyEvents
        </p>
      </div>
    </article>
  </article>
</DocumentFragment>
`;
